experimental/ir/lower_resolve.go

```diff
@@ -123,6 +119,36 @@ func resolveFieldType(field Member, r *report.Report) {
 	if sym.Kind().IsType() {
 		field.raw.elem.file = sym.ref.file
 		field.raw.elem.ptr = arena.Pointer[rawType](sym.raw.data)
+
+		if mf := sym.AsType().MapField(); !mf.IsZero() {
+			r.Errorf("use of synthetic map entry type").Apply(
+				report.Snippetf(path, "referenced here"),
+				report.Snippetf(mf.TypeAST(), "synthesized by this type"),
+				report.Helpf("despite having a user-visible symbol, map entry "+
+					"types cannot be used as field types"),
+			)
+		}
+
+		if !field.Container().MapField().IsZero() && field.Number() == 1 {
+			// Legalize that the key type must be comparable.
+			ty := sym.AsType()
+			if !ty.Predeclared().IsMapKey() {
+				d := r.Error(errTypeCheck{
+					want: "comparable type",
+					got:  ty,
+					expr: field.TypeAST(),
+				}).Apply(report.Helpf(
+					"comparable types are integer types, " +
+						"`string`, and `bool`",
```

Comment:
I don't love the term "comparable" here since that notion is really target-language dependent so could be confusing here. For example, enums are comparable in most target languages yet conspicuously absent from this set of allowed types. (Though bytes are not comparable in Go, I think they are actually comparable in most target languages, too.)
```suggestion
					want: "allowed map key type",
					got:  ty,
					expr: field.TypeAST(),
				}).Apply(report.Helpf(
					"allowed map key types include only integer types, " +
						"`string`, and `bool`",
```